File: CSharp/226-InvertBinaryTree/226InvertBinaryTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
///////////////////////////////////////////////////////
// RECURSION - Depth First Search (DFS)

// Time Complexity: O(n)
//// You must visit each node for the swap.

// Space Complexity: O(H)
/* 
H is the height of the binary tree. The recursion stack will hold at most 
H function calls at any given time. In worst case for a skewed tree
Like a linked list H will be n giving us O(n)
Best case balanced binary tree h will be log(n) so we get O(log n)
*/
public class Solution {
    public TreeNode InvertTree(TreeNode root) {
        if (root == null) return null;

        TreeNode temp = root.left;
        root.left = InvertTree(root.right);
        root.right = InvertTree(temp);
        return root;
    }
}

File: Javascript/226-InvertBinaryTree/226InvertBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
 
 // Depth First Search, Destructuring assignment, and recursion 
 // Time Complexity: O(n) Where n is the number of nodes in the tree. We must traverse the entire tree to invert it.
 // Space Complexity: O(h) Each recursive call adds a new frame to the call stack, which can lead to O(h) space complexity, where h is the height of the tree. In the worst case (for a skewed tree), this could be O(n). For a balanced tree, it would be O(log n).

 var invertTree = function(root){
    if (!root) return null;
    [root.left, root.right] = [invertTree(root.right), invertTree(root.left)];
    return root;
}